jib-core/src/main/java/com/google/cloud/tools/jib/image/ReproducibleLayerBuilder.java

```diff
@@ -101,15 +154,35 @@ public ReproducibleLayerBuilder addFiles(List<Path> sourceFiles, String extracti
    * @throws IOException if walking the source files fails
    */
   public UnwrittenLayer build() throws IOException {
-    List<TarArchiveEntry> filesystemEntries = new ArrayList<>();
+    LinkedHashSet<CustomHashWrapper<TarArchiveEntry, String>> filesystemEntrySet =
+        new LinkedHashSet<>();
 
     // Adds all the layer entries as tar entries.
-    for (LayerEntry layerEntry : layerEntries.build()) {
-      filesystemEntries.addAll(buildAsTarArchiveEntries(layerEntry));
+    List<LayerEntry> layerEntries = this.layerEntries.build();
+    for (LayerEntry layerEntry : layerEntries) {
+      // Converts layerEntry to list of TarArchiveEntrys.
+      List<TarArchiveEntry> tarArchiveEntries = buildAsTarArchiveEntries(layerEntry);
+
+      for (TarArchiveEntry tarArchiveEntry : tarArchiveEntries) {
+        // Adds all directories along extraction paths to explicitly set permissions for those
+        // directories.
+        for (Path parentDirectory : getParentDirectories(Paths.get(tarArchiveEntry.getName()))) {
```

Comment:
I wonder if this wouldn't be better in TarStreamBuilder.  This approach also generates a lot of garbage since for only the first file per parent actually needs to walk the parents.

An alternative would be to first sort the file list, then walk the list and add directory elements all at once to the TarStreamBuilder, and then add the files.